collaboration/extending-kyma-cli.md

```diff
@@ -0,0 +1,351 @@
+
+Example structure of config.yaml
+
+```yaml
+name: <function_name>
+namespace: <namespace> #default by default
+events:
+  - event1
+  - event2
+environment-vars:
+  - foo:bar
+api:
+  url: <url>
+  actions:
+    - GET
+    - POST
+  security:
+    - JWT
+replicas: 1
+memory:
+  min:
+  max:
+cpu:
+  min:
+  max:
+runtime:
+debug-string:
+```
+
+##### Template files
+
+###### <function-code.extension>
+
+  Out of the box from the init command we should give an easy example which can be extended by the developer later. The example should be possible to be deployed directly on the kyma cluster. Here below there is a node example
+
+  ```js
+    module.exports = {
+      main: function (event, context) {
+        return 'Hello World';
+      }
+    };
+  ```
+
+###### <dependecies.extension>
+
+We should generate an empty package.json file which can be edited by developer as and when needed.
+
+###### index.js
+
+We should provide `index.js`(which would start the server and serve the handler function) out of the box which can be used by the developer to test his code locally.
+
+###### readme
+
+We should also provide a readme.md which would consist of following instructions:
+
+* How he can deploy the function into the cluster.
+* How he can run the function locally.
+* How to use the `config.yaml` file
+* Folder structure.
+* How to test debug it locally like using telepresence.
+
+##### Testing/Debugging
+
+*Local*
+
+We can introduce commands to test the function locally and also if the developer wants to debug the code. These can be achieved through make file.
+
+*Using Events from remote cluster*
+
+Sometimes the event payloads can be complex to be created locally. For such cases its nice to have the events on remote cluster triggering the code on the workstation. This would enable the developer to debug his code on the workstation using events being sent to the remote cluster. We can introduce command for example below:
+
+`kyma function debug <function-name> -n <namespace>`
+
+> We can use telepresence for local debugging. We can either use --swap-deployment or --new-deployment option.
+
+Telepresence would create a proxy in the kyma cluster and redirect the calls to local codebase. One can attach the debugger to his code base along with breakpoints to debug the code.
+
+##### Deploying and running the project
+
+The function generated by `init` command should be deployable onto kyma cluster. For deploying the file onto kyma cluster it would read the values from `config.yaml`. If the api name is not set then we should show some warning message and deploy.
+
+For running/debugging it locally we should provide makefile. The running/debugging commands should be part of readme.
+
+#### Working with deployed kyma cluster
+
+We should allow deployment of the functions on a k8s cluster. Following commands should be supported
+
+##### CRUD
+
+##### Commands that use config
+
+`kyma function create -p <path-to-function-dir>`
```

Comment:
Why do you think it is important to distinguish between "create" and "update" here? Shouldn't a `kubectl apply` way of deploying the function to the cluster be sufficient? Otherwise the developer (or CI system) needs to build logic to check if it is existing and so on.

Reply:
From a naming perspective i think terms like `apply` or `deploy` or possibly `install` describe more explicitly that this command will transfer the local state to the cluster rather than `create`/`update`.

Reply:
Yeah apply sounds better here